internal/util/invalid_serialization_detector.go

```diff
@@ -6,27 +6,39 @@ import (
 	"reflect"
 )
 
+// InvalidSerializationDetectorOps is a list of operations in the order that they
+// are performed by this schema.
+//
+// NOTE:  Because this list is global, it will only work properly if the test
+// uses only a single instance of this schema.
+var InvalidSerializationDetectorOps []string
+
 // NewInvalidSerializationDetectorSchema creates a new test schema type.
 func NewInvalidSerializationDetectorSchema() *InvalidSerializationDetectorSchema {
+	// Because of the use of a global variable for visibility, we can only permit
+	// a single instance of this type.
+	if len(InvalidSerializationDetectorOps) > 0 {
+		panic("Multiple instances of InvalidSerializationDetectorSchema detected")
+	}
+	InvalidSerializationDetectorOps = []string{"init"}
 	return &InvalidSerializationDetectorSchema{}
 }
 
-// InvalidSerializationDetectorSchema is a testing type that detects double serialization or double unserialization,
-// which helps prevent bugs.
-// It is backed by a string, which is set to the last operation performed. If the last operation matches the
-// current operation, then that indicates double serialization or double unserialization.
+// InvalidSerializationDetectorSchema is a testing type that detects double
+// serialization or double unserialization which could result in corrupted data.
+// The schema tracks the sequence of operations which are performed. The last
```

Comment:
Although the previous operation will currently be the last operation, in the context of a sequence I'd prefer the word "previous" over "last".

Reply:
Sadly, the notion of a sequenced list of operations turned out to be impractical:  it is not unreasonable for the calling code to "validate" the data and then serialize it (or to serialize two values from separate fields which use the same schema), which results in there being two serialization operations in a row.  (The engine doesn't actually do this, but Jared wrote a full suite of tests for the test-schema, and one of them does the equivalent.)

Matt and I discussed and came to the conclusion that the only realistic way to detect a double operation is via the data, as Jared had done originally.  That is sufficient to detect _double_ operations; so, what I've now added is visibility which allows us to detect _single_ operations.  That is, if **no** serialization or unserialization occurs, then it is trivial to claim that no _double-operations_ occurred; so, after showing that no double-ops occurred, the test needs to confirm that at least one single-op occurred, which it does, now (using a _counter_ as Matt original suggested).  😅

Reply:
Rats:  I need to update this comment, now!